src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -70,14 +70,16 @@ public virtual IDisposable Start()
             var server = ServerFactory.Start(_serverInstance,
                 async features =>
                 {
-                    var httpContext = contextFactory.CreateHttpContext(features);
-                    httpContext.ApplicationServices = _applicationServices;
-                    var requestIdentifier = GetRequestIdentifier(httpContext);
-
-                    using (logger.BeginScope("Request Id: {RequestId}", requestIdentifier))
+                    using (var httpContext = contextFactory.CreateHttpContext(features))
```

Comment:
The scope of HttpContext is not this deterministic. IHttpResponseFeature has two events, OnStarting and OnCompleted, that may not have fired before the end of this using block. Both of these events may have user state that contains the HttpContext.  I'm not sure there is a simple workaround.